client/verta/verta/dataset/_s3.py

```diff
@@ -72,6 +79,40 @@ def __init__(self, paths):
 
         s3_metadata = six.viewvalues(obj_paths_to_metadata)
         self._msg.s3.components.extend(s3_metadata)  # pylint: disable=no-member
+        self._components_to_upload = dict()
+
+        # TODO: maybe make this happen in the above loop to help avoid race condition
+        # TODO: or it should happen in commit.save() so that's it's closer to upload time
```

Comment:
I think putting it closer to save makes sense. Otherwise you're downloading things that you might not use.